### PR TITLE
[Property Proxy Typescript Migration 3/3] migrating property proxy to TS

### PR DESCRIPTION
**Pull request by karlbom**
this is a WIP and has most of the changes already made, it still needs to cleaned up

should only be merged and reviewed once it is out of "draft" status and the PRs #6740  and #6741  have been merged successfully

### CODE REVIEW

**Comment by karlbom** [comment]
I added some todos for areas that still need work but this PR moves the property-proxy to typescript.

but the changes should be ready to review

File: experimental/PropertyDDS/packages/property-proxy/.eslintrc.js [hunk 1]
```diff
@@ -3,17 +3,20 @@
...
-const eslintConfig = require("@fluidframework/eslint-config-fluid");
+module.exports = {
+    "extends": [
+        "@fluidframework/eslint-config-fluid/eslint7"
+    ],
+    "parserOptions": {
+        "project": ["./tsconfig.json", "./src/test/tsconfig.json"]
+    },
+    "rules": {
+        // TODO(marcus): remove the linting issues
```
Comment (karlbom) [inline experimental/PropertyDDS/packages/property-proxy/.eslintrc.js:14]: I changed the eslint configuration to the same way other packages define their linting, I had to deactivate some of the linting rules for now since fixing them in this PR will be a rather large change that is probably better to be done in a separate PR

File: experimental/PropertyDDS/packages/property-proxy/src/componentArray.ts [hunk 2]
```diff
@@ -128,49 +133,43 @@ class ComponentArray extends Array {
     /**
      * @inheritdoc
      */
-    lastIndexOf(searchElement, fromIndex) {
-        // Only handle non-primitive cases,
-        // primitive cases can be handled implicitly (see componentArrayProxyHandler.js)
-
-        if (PropertyFactory.instanceOf(!searchElement, 'BaseProperty')) {
```
Comment (karlbom) [inline experimental/PropertyDDS/packages/property-proxy/src/componentArray.ts:135]: this branch should just never be taken !searchElement coearces it to a boolean so will either be false or true(if false was the provided element) neither of the cases will ever be an instance of `BaseProperty` therefor this if branch will never be taken

File: experimental/PropertyDDS/packages/property-proxy/src/componentSet.ts [hunk 3]
```diff
@@ -105,10 +113,15 @@ class ComponentSet extends Set {
      */
     entries() {
         const keys = this.property.getIds();
-        const entriesIterator = function*() {
-            for (let i = 0; i < keys.length; i++) {
-                const proxy = PropertyProxy.proxify(this.property.get(keys[i]));
-                yield [proxy, proxy];
+        const entriesIterator = function*(this: ComponentSet): Generator<[any, any]> {
+            for (const key of keys) {
+                const property_to_proxy = this.property.get(key);
```
Comment (karlbom) [inline experimental/PropertyDDS/packages/property-proxy/src/componentSet.ts:118]: explicitly throw if this is `undefined`, this should not happen

File: experimental/PropertyDDS/packages/property-proxy/src/componentSet.ts [hunk 4]
```diff
@@ -118,18 +131,23 @@ class ComponentSet extends Set {
...
      */
-    forEach(func) {
+    forEach(func: (value, key, set) => void) {
         const keys = this.property.getIds();
-        for (let i = 0; i < keys.length; i++) {
-            const value = PropertyProxy.proxify(this.property.get(keys[i]));
-            func(value, value, this);
+        for (const key of keys) {
+            const property_to_proxy = this.property.get(key);
+            if (property_to_proxy) {
```
Comment (karlbom) [inline experimental/PropertyDDS/packages/property-proxy/src/componentSet.ts:138]: same as above

File: experimental/PropertyDDS/packages/property-proxy/src/errors.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
+/*!
```
Comment (karlbom) [inline experimental/PropertyDDS/packages/property-proxy/src/errors.ts:1]: I wonder why some renamings are tracked in the git view as a rename from js to ts with some changes in the file, while others are considered a remove of the js file with an addition of a new ts files were renamed the same way and changes were done in a separate commit.

File: experimental/PropertyDDS/packages/property-proxy/src/arrayProxyHandler.ts [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+     * The set trap handles setting of properties. If key is a number >= 0 it sets the
+     * property at that index in the {@link external:ArrayProperty ArrayProperty}.
+     * If the key is 'length' it sets a new length for the {@link external:ArrayProperty ArrayProperty}.
+     * Otherwise, it just sets it on the associated {@link ComponentArray}.
+     * @param target The {@link ComponentArray} the Proxy handles.
+     * @param key The name of the property/function that is to be accessed.
+     * @param value The value to be set.
+     * @return Returns a boolean.
+     */
+    set(target: ComponentArray, key: string | number, value: any) {
```
Comment (karlbom) [inline experimental/PropertyDDS/packages/property-proxy/src/arrayProxyHandler.ts:180]: this method was refactored and shot be compared agains the old one

**Comment by karlbom** [comment]
@DLehenbauer

**Review by nedalhy — APPROVED** [review]

**Comment by nedalhy** [comment]
@DLehenbauer can you please merge this one ?

**Review by DLehenbauer — APPROVED** [review]
